libraries/abstractions/platform/freertos/iot_network_udp.c

```diff
@@ -0,0 +1,543 @@
+    {
+        IotLogError( "Failed to resolve DNS address %s.",
+                     pServerInfo->pHostName );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_NETWORK_FAILURE );
+    }
+
+    *pNetworkConnection = pvPortMalloc( sizeof( _networkConnection_t ) );
+
+    if( *pNetworkConnection == NULL )
+    {
+        IotLogError( "Failed to allocate memory for new network connection." );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_NETWORK_NO_MEMORY );
+    }
+
+    /* Clear the connection information. */
+    ( void ) memset( *pNetworkConnection, 0x00, sizeof( _networkConnection_t ) );
+
+    ( *pNetworkConnection )->socket = udpSocket;
+    ( *pNetworkConnection )->serverAddr.ulAddress = udpServerAddr;
+    ( *pNetworkConnection )->serverAddr.usPort = SOCKETS_htons( pServerInfo->port );
+    ( *pNetworkConnection )->serverAddr.ucSocketDomain = SOCKETS_AF_INET;
+    ( *pNetworkConnection )->strictServerAddrCheck = pServerInfo->strictServerAddrCheck;
+
+    /* Create the connection event flags and mutex. */
+    pConnectionFlags = xEventGroupCreateStatic( &( ( *pNetworkConnection )->connectionFlags ) );
+
+    /* Static event flags and mutex creation should never fail. The handles
+     * should point inside the connection object. */
+    configASSERT( pConnectionFlags == ( EventGroupHandle_t ) &( ( *pNetworkConnection )->connectionFlags ) );
+
+
+    IOT_FUNCTION_CLEANUP_BEGIN();
+
+    if( status != IOT_NETWORK_SUCCESS )
+    {
+        if( udpSocket != SOCKETS_INVALID_SOCKET )
+        {
+            SOCKETS_Close( udpSocket );
+        }
+
+        /* Clear the connection information. */
+        if( ( pNetworkConnection != NULL ) && ( ( *pNetworkConnection ) != NULL ) )
+        {
+            vPortFree( *pNetworkConnection );
+        }
+    }
+
+    IOT_FUNCTION_CLEANUP_END();
+}
+/*-----------------------------------------------------------*/
+
+IotNetworkError_t IotNetworkUdp_SetReceiveCallback( void * pConnection,
+                                                    IotNetworkReceiveCallback_t receiveCallback,
+                                                    void * pContext )
+{
+    IOT_FUNCTION_ENTRY( IotNetworkError_t, IOT_NETWORK_SUCCESS );
+
+    /* Cast network connection to the correct type. */
+    _networkConnection_t * pNetworkConnection = ( _networkConnection_t * ) pConnection;
+
+    if( pConnection == NULL )
+    {
+        IotLogWarn( "Invalid connection parameter" );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_NETWORK_BAD_PARAMETER );
+    }
+
+    /* Set the receive callback and context. Each pConnection has its own receiveCallback and pReceiveContext. */
+    pNetworkConnection->receiveCallback = receiveCallback;
+    pNetworkConnection->pReceiveContext = pContext;
+
+    /* No flags should be set. */
+    configASSERT( xEventGroupGetBits( ( EventGroupHandle_t ) &( pNetworkConnection->connectionFlags ) ) == 0 );
+
+    /* Create task that waits for incoming data. */
+    if( xTaskCreate( _networkReceiveTask,
```

Comment:
What do we think is the use case in which an app will be using UDP and want us to incur the overhead of an additional task for handling async receives?